proto/message_api/v1/authn.proto

```diff
@@ -0,0 +1,66 @@
+// Authn protocol
+syntax = "proto3";
+package xmtp.message_api.v1;
+
+option go_package = "./message_api/v1";
+
+// ----------------------------------------------------------------------------
+// These objects are copied from xmtp-js. 
+// Differences:
+// * PublicKey.signature is made non-optional
+// ----------------------------------------------------------------------------
+
+// Signature represents a generalized public key signature,
+// defined as a union to support cryptographic algorithm agility.
+message Signature {
+    // ECDSA signature in its compact format
+    message ECDSACompact {
+        bytes bytes = 1; // compact representation [ R || S ], 64 bytes
+        uint32 recovery = 2; // recovery bit
+    }
+    oneof union {
+        ECDSACompact ecdsa_compact = 1;
+    }
+}
+
+// Secp256k1 public key in uncompressed point format.
+message Secp256k1Uncompresed {
+    // uncompressed point with prefix (0x04) [ P || X || Y ], 65 bytes
+    bytes bytes = 1; 
+}
+
+// PublicKey represents a generalized public key,
+// defined as a union to support cryptographic algorithm agility.
+message PublicKey {
+    uint64 timestamp = 1;
+    Signature signature = 2;  // signature is optional in the xmtp-js version
+    oneof union {
+        Secp256k1Uncompresed secp256k1_uncompressed = 3;
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+// Token is used by clients to prove to the nodes
+// that they are serving a specific wallet.
+message Token {
+    // identity key signed by a wallet
+    PublicKey identity_key = 1; 
+    // encoded bytes of AuthData
+    bytes auth_data_bytes = 2;
+    // identity key signature of AuthData bytes
+    Signature auth_data_signature = 3; 
+}
+
+// AuthData carries token parameters that are authenticated
+// by the identity key signature.
+// It is embedded in the Token structure as bytes
+// so that the bytes don't need to be reconstructed
+// to verify the token signature.
+message AuthData {
+    // address of the wallet
+    string wallet_addr = 1;
+    // time when the token was generated/signed 
+    uint64 timestamp = 2; 
```

Comment:
One thought: Should we turn this into `valid_until` to allow the client to control the validity expiration within node enforced constraint? (e.g. let's say the nodes will only accept up to an hour or something).

Reply:
It's hard for me to imagine a client developer choosing anything less than the maximum amount of time, given that the tokens are there to protect XMTP more than they are to protect the user.

Reply:
I look at it differently. A stolen token is not an issue for the network, it doesn't matter who used up the limit. It is an issue for the user, since it allows someone to DOS them. So I was thinking a more paranoid user may want to reduce the expiration period at the cost of having to create tokens more often.

Reply:
I still don't see it. If we have a problem where people are able to compromise these tokens and use them for DOS, pretty sure the response is going to be to not use the network at all. Being DOS'd for 20 minutes and 60 feels about the same to the end user: unacceptable.

Reply:
Fair enough, I'll leave this be.

Reply:
I think having a time the token was generated would be helpful into the future.

Though I could see having an addition field for expiry like in the JWT spec.

Reply:
If we care about having the client be aware of when the tokens are due to expire we could always add a response header where the server would provide that value. Not sure how useful it is given that the client can always just regenerate when a request fails with a 401 status code, but that feels preferable to having the client guess when the token is going to expire.